src/v/crypto/ossl_context_service.cc

```diff
@@ -0,0 +1,288 @@
+    impl& operator=(impl&&) noexcept = delete;
+
+    ss::future<> start() {
+        lg.trace("Starting service...");
+        if (ss::this_shard_id() == 0) {
+            // On main shard, load the 'null' provider to the default context
+            // This prevents the default context from performing any
+            // cryptographic operation
+            _defctxnull = internal::OSSL_PROVIDER_ptr(
+              OSSL_PROVIDER_load(nullptr, "null"));
+            lg.debug("Loaded null into global provider");
+            // We also need to create a library contxt and load it in just
+            // within the thread worker's thread.  This is so the krb5 library
+            // will use this loaded context for any operations it needs to
+            // perform
+            _thread_worker_ctx = internal::OSSL_LIB_CTX_ptr(OSSL_LIB_CTX_new());
+            auto init_resp = co_await _thread_worker.submit([this] {
+                return initialize_worker_thread(
+                  _thread_worker_ctx.get(),
+                  _module_path,
+                  _config_file,
+                  _fips_mode);
+            });
+
+            if (init_resp.has_error()) {
+                throw exception(init_resp.assume_error());
+            }
+
+            _initialize_thread_worker_holder = std::move(
+              init_resp.assume_value());
+        }
+
+        _shard_ctx = internal::OSSL_LIB_CTX_ptr(OSSL_LIB_CTX_new());
+        _old_context = OSSL_LIB_CTX_set0_default(_shard_ctx.get());
+        lg.debug(
+          "Created new shard context for {} replacing {}",
+          fmt::ptr(_shard_ctx.get()),
+          fmt::ptr(_old_context));
+        auto init_resp = co_await _thread_worker.submit([this] {
+            return initialize_openssl(
+              _shard_ctx.get(), _module_path, _config_file, _fips_mode);
+        });
+
+        if (init_resp.has_failure()) {
+            throw exception(init_resp.assume_error());
+        }
+
+        _crypto_provider = std::move(init_resp.assume_value().crypto_provider);
+        _base_provider = std::move(init_resp.assume_value().base_provider);
+    }
+
+    ss::future<> stop() {
+        lg.trace("Stopping service...");
+        _base_provider.reset();
+        _crypto_provider.reset();
+        if (_old_context != nullptr) {
+            auto replaced_context = OSSL_LIB_CTX_set0_default(_old_context);
+            lg.debug(
+              "Reverted to old context {} and received back {} (expecting {})",
+              fmt::ptr(_old_context),
+              fmt::ptr(replaced_context),
+              fmt::ptr(_shard_ctx.get()));
+            vassert(
+              replaced_context == _shard_ctx.get(),
+              "Replacing original context returns unexpected library context");
+        } else {
+            lg.warn("Original context is null... startup failed?");
+        }
+        _old_context = nullptr;
+        _shard_ctx.reset();
+        _shard_ctx = nullptr;
```

Comment:
FWIW these two lines do the same thing

Reply:
whoops meant to replace 228 with 227 and not have both :)